[harithh07] iP

harithh07 opened the pull request:
### Rasputin

> “Your mind is for having ideas, not holding them.” – David Allen [(source)](https://dansilvestre.com/productivity-quotes/)

Rasputin is the only reminder app you'll need. It's

- Text-based
- Easy to learn
- ~~FAST~~ _SUPER_ FAST to use

Want to use it? To start, 

1. Download the file [here]([rasputin.jar](https://github.com/harithh07/ip/releases/download/A-Jar/rasputin.jar))
2. Run your preferred terminal
3. Navigate to the directory that you downloaded the .jar file
4. Type `java -jar rasputin.jar`

Best of all, it's **FREE!** :tada:

Features: 

- [x] Managing tasks
- [x] GUI
- [ ] Reminders (coming soon)

---

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
Rasputin rasputin = new Rasputin("src/main/data/rasputin.txt");
rasputin.run();
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

harithh07 committed:
Added greeting and exit message

harithh07 committed:
Added echo function

harithh07 committed:
Added list to store user input text

harithh07 committed:
Added ability to display list

harithh07 committed:
Add the ability to mark tasks as done

harithh07 committed:
Added Todo task class

harithh07 committed:
Add Deadline task class

harithh07 committed:
Add Event task class

harithh07 committed:
Add count of list after adding task

harithh07 committed:
Add automated text UI testing

harithh07 committed:
Add exceptions

harithh07 committed:
Add ability to delete tasks

harithh07 committed:
Fixed input check

harithh07 committed:
Add reading task list from text file

harithh07 committed:
Add close FileReader and BufferedReader

harithh07 committed:
Add reading isDone for tasks

harithh07 committed:
Add writing to text file at the end of program

harithh07 committed:
Fix text-ui-test to accommodate reading and writing text file

harithh07 committed:
Add input for text-ui-test

harithh07 committed:
Change if else blocks in Rasputin.java to switch case block

harithh07 committed:
Add dates and time to Deadline task

harithh07 committed:
Fix printing date and time with DateTimeFormatter

harithh07 committed:
Add dates and time to Event task

harithh07 committed:
Encapsulate Storage class for reading and writing files

harithh07 committed:
Encapsulate TaskList class for storing list of tasks

harithh07 committed:
Encapsulate Ui class

harithh07 committed:
Encapsulate classes for Command, Parser and Ui and add exceptions

harithh07 committed:
Merge branch 'branch-MoreOOP'

* branch-MoreOOP:
  Encapsulate classes for Command, Parser and Ui and add exceptions
  Encapsulate Ui class
  Encapsulate TaskList class for storing list of tasks
  Encapsulate Storage class for reading and writing files

harithh07 committed:
Package classes

harithh07 committed:
Merge commit '81a9c537afe8492675eb253968939364d65db091'

* commit '81a9c537afe8492675eb253968939364d65db091':
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

harithh07 committed:
Add gradle

harithh07 committed:
Add JUnit tests

harithh07 committed:
Add JavaDoc comments

harithh07 committed:
Change code to comply to coding standard

harithh07 committed:
Add command to find tasks with a keyword

harithh07 committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add JavaDoc comments

harithh07 committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Change code to comply to coding standard

harithh07 committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add command to find tasks with a keyword

# Conflicts:
#	src/main/java/Task/TaskList.java

harithh07 committed:
Add checkstyle

harithh07 committed:
Update package names

harithh07 committed:
Add GUI with JavaFX

harithh07 committed:
Clean up redundant code

harithh07 committed:
Add minor cosmetic changes

harithh07 committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Add minor cosmetic changes
  Clean up redundant code
  Add GUI with JavaFX
  Update package names

celeschai commented on src/main/java/rasputin/task/Deadline.java:
```diff
@@ -0,0 +1,61 @@
+    String by;
+    LocalDate date;
+    LocalDateTime dateTime;
+    boolean hasTime;
```
Should it be better to name `hasTime` as `isTime`?

celeschai commented on src/main/java/rasputin/task/TaskList.java:
```diff
@@ -0,0 +1,113 @@
+     * @param index of task to be retrieved.
+     * @return Task
+     */
+    public Task get(int index) {
```
Would it be clearer to use a clearer name that shows what are we getting?

celeschai commented on src/main/java/rasputin/task/TaskList.java:
```diff
@@ -0,0 +1,113 @@
+     *
+     * @return Size of list.
+     */
+    public int size() {
```
Should this be named `getSize`?

celeschai commented on src/test/java/storage/StorageTest.java:
```diff
@@ -0,0 +1,34 @@
+public class StorageTest {
+
+    @Test
+    public void testReadFile() {
+        Storage storage = new Storage("src/main/data/rasputin.txt");
+        ArrayList<Task> tasks = new ArrayList<>();
+        Deadline deadline1 = new Deadline("return book", "2024-08-29 1800");
+        Event event = new Event("project meeting","2024-08-28 1400", "2024-08-28 1500");
+        event.markAsDone();
+        Todo todo = new Todo("read book");
+        todo.markAsDone();
+        Deadline deadline2 = new Deadline("update calendar", "2024-08-30");
+        tasks.add(deadline1);
+        tasks.add(event);
+        tasks.add(todo);
+        tasks.add(deadline2);
+        assertEquals(tasks.toString(), storage.readFile().toString());
```
Perhaps breaking down the code a bit with some comments will help another reader understand the testing being done here?

celeschai commented on src/main/java/rasputin/gui/DialogBox.java:
```diff
@@ -0,0 +1,68 @@
+package rasputin.gui;
```
I like that you separated the GUI code from the main part!

celeschai commented on src/main/java/rasputin/command/InvalidCommand.java:
```diff
@@ -0,0 +1,33 @@
+ */
+public class InvalidCommand extends Command {
+
+    private String message;
```
Maybe the name of the message can be more specific so readers can know what message we are referring to?

celeschai reviewed (commented):
Overall I think you did a really good job!
JavaDocs are in place and you managed to follow the naming conventions most of the time, even using plurals for all lists, which I personally missed.

harithh07 committed:
Fix gui bug

jeanyang04 commented on src/main/java/rasputin/command/AddCommand.java:
```diff
@@ -0,0 +1,100 @@
+                desc = deadlineSplit[0];
+                if (deadlineSplit.length < 2) {
+                    throw new InvalidTaskException("ERROR! Deadline tasks require a deadline to be completed by.");
+                }
+                String by = deadlineSplit[1];
+                Deadline deadline = null;
+                try {
+                    deadline = new Deadline(desc, by);
+                } catch (DateTimeParseException e) {
+                    throw new InvalidTaskException("ERROR! Invalid deadline format.");
+                }
+                tasks.add(deadline);
+                return Ui.printAddTask(deadline, tasks);
+            case "event":
+                str = input.substring(6);
+                String[] durations = str.split(" /from ");
+                desc = durations[0];
+                if (durations.length < 2) {
+                    throw new InvalidTaskException("ERROR! Event tasks require a duration for the event.");
+                }
+                String fromSplit = durations[1];
+                String[] toSplit = fromSplit.split(" /to ");
+                if (toSplit.length < 2) {
+                    throw new InvalidTaskException("ERROR! Event tasks require a duration for the event.");
+                }
+                String from = toSplit[0];
+                String to = toSplit[1];
+
+                Event event = null;
+                try {
+                    event = new Event(desc, from, to);
+                } catch (DateTimeParseException e) {
+                    throw new InvalidTaskException("ERROR! Invalid event duration format.");
+                }
+                tasks.add(event);
+                return Ui.printAddTask(event, tasks);
+            default:
+                throw new InvalidTaskException("ERROR! Invalid task.");
+        }
+    }
```
Perhaps the indentation can be changed since cases within a switch block should not be indented

jeanyang04 commented on src/main/java/rasputin/command/ExitCommand.java:
```diff
@@ -0,0 +1,46 @@
+package rasputin.command;
+
+import javafx.application.Platform;
+import rasputin.task.TaskList;
+
+import rasputin.storage.Storage;
+
+import rasputin.gui.Ui;
```
Perhaps the rasputin imports can be put together, while the javafx import can be put on top for better readability?

jeanyang04 commented on src/main/java/rasputin/task/Event.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    @Override
+    public String getType() {
+        return "Event";
+    }
```
Consider adding javadocs to explain the functionality of this method

jeanyang04 reviewed (commented):
Overall I liked how you put everything into their own classes/packages, which was very neat. However do note that there a a few nits to fix. :D

harithh07 committed:
Extract out methods within writeFile() in Storage class

The writeFile() method in the Storage class was too long with multiple
switch cases to write to the file in the case of the different tasks.

Extracting out a method to get the format in which the task should
be written to the file improves the code quality.

Let's extract the switch case code block into a method in the respective
classes to return a String in the proper format to be written to the
task file.

harithh07 committed:
Delete redundant code

harithh07 committed:
Storage readFile(): specifying the file path with a proper relative path

readFile() takes in a file path that will not exist when running the
chatbot outside of the IDE.

Let's change the file path taken in to be a relative file path and
create a new directory to store the file if it does not exist.

harithh07 committed:
Add assertions

harithh07 committed:
Merge pull request #1 from harithh07/branch-A-Assertions

Add assertions

harithh07 committed:
Merge branch 'master' of https://github.com/harithh07/ip

* 'master' of https://github.com/harithh07/ip:
  Add assertions

harithh07 committed:
Merge pull request #2 from harithh07/branch-A-CodeQuality

Branch A-CodeQuality

harithh07 committed:
Add undo command

harithh07 committed:
Add undo command

harithh07 committed:
Merge branch 'branch-C-Undo'

* branch-C-Undo:
  Add undo command
  Add undo command

harithh07 committed:
Add and edit JavaDocs

harithh07 committed:
Change pictures for user and Rasputin

harithh07 committed:
Change picture for Rasputin and change Rasputin's responses

harithh07 committed:
Merge branch 'branch-A-Personality'

* branch-A-Personality:
  Change picture for Rasputin and change Rasputin's responses
  Change pictures for user and Rasputin

harithh07 committed:
Increase height of window

harithh07 committed:
Fix sendButton position

harithh07 committed:
Add product screenshot

harithh07 committed:
Update user guide

harithh07 committed:
Merge branch 'branch-A-UserGuide'

* branch-A-UserGuide:
  Update user guide
  Add product screenshot
  Fix sendButton position
  Increase height of window

harithh07 committed:
Fix formatting issues in README

harithh07 committed:
Fix formatting in README

harithh07 committed:
Fix GUI for Windows

harithh07 committed:
Fix jUnit test StorageTest

harithh07 committed:
Add credits to logos for Rasputin and User
